### PR TITLE
Features/cit 354/c8y serializer

### PR DESCRIPTION
**Pull request by PradeepKiruvale**
## Proposed changes

This PR captures the implementation of c8yserializer.

## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x ] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x ] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [ x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: mapper/cumulocity/c8y_translator_lib/src/json.rs [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+                    serializer.start_group(&measurement.name)?;
+                    for s in measurement.values.iter() {
+                        serializer.measurement(&s.name, s.value)?;
+                    }
+                    serializer.end_group()?;
+                }
+            }
+        }
+        Ok(serializer.bytes()?)
+    }
```
Comment (makr11st) [inline mapper/cumulocity/c8y_translator_lib/src/json.rs:60]: These are associated method but couldn't they just be module functions?

File: mapper/cumulocity/c8y_translator_lib/src/json.rs [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+
+use crate::serializer;
+use chrono::prelude::*;
+use thin_edge_json::{
+    json::{ThinEdgeJson, ThinEdgeValue},
+    measurement::GroupedMeasurementVisitor,
+};
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct CumulocityJson;
```
Comment (makr11st) [inline mapper/cumulocity/c8y_translator_lib/src/json.rs:23]: What's the purpose of this struct here? Is it only to hold the associated functions below? I would remove it and make these methods just module functions.

File: mapper/cumulocity/c8y_translator_lib/src/json.rs [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+
+impl CumulocityJson {
+    ///Convert from thinedgejson to c8y_json
+    pub fn from_thin_edge_json(
+        input: &[u8],
+    ) -> Result<Vec<u8>, serializer::C8yJsonSerializationError> {
+        let local_time_now: DateTime<Local> = Local::now();
+        let timestamp = local_time_now.with_timezone(local_time_now.offset());
```
Comment (makr11st) [inline mapper/cumulocity/c8y_translator_lib/src/json.rs:31]: Is that right to generate the timestamp at the time of serialisation and not time of arrival? Is there any requirement when should it be generated?

File: mapper/cumulocity/c8y_translator_lib/src/json.rs [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+        }
+        Ok(serializer.bytes()?)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn test_timestamp() -> DateTime<FixedOffset> {
```
Comment (makr11st) [inline mapper/cumulocity/c8y_translator_lib/src/json.rs:67]:
```suggestion
    fn generate_test_timestamp() -> DateTime<FixedOffset> {
```

File: mapper/cumulocity/c8y_translator_lib/src/serializer.rs [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    #[error("Unexpected end of data")]
+    UnexpectedEndOfData,
+
+    #[error("Unexpected end of group")]
+    UnexpectedEndOfGroup,
+
+    #[error("Unexpected start of group")]
+    UnexpectedStartOfGroup,
+}
+
```
Comment (makr11st) [inline mapper/cumulocity/c8y_translator_lib/src/serializer.rs:37]: Why do you need to have 2 separate error types here? Can't you combine them into one and use as such?

File: mapper/cumulocity/c8y_translator_lib/src/serializer.rs [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    IoError(#[from] std::io::Error),
+
+    #[error(transparent)]
+    MeasurementCollectorError(#[from] MeasurementStreamError),
+
+    #[error(transparent)]
+    ThinEdgeJsonParseError(#[from] ThinEdgeJsonError),
+}
+#[derive(thiserror::Error, Debug, Eq, PartialEq)]
+pub enum MeasurementStreamError {
```
Comment (makr11st) [inline mapper/cumulocity/c8y_translator_lib/src/serializer.rs:24]: Is that name actually right, what kind of stream do we have here?

File: mapper/cumulocity/c8y_translator_lib/src/serializer.rs [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    UnexpectedEndOfData,
+
+    #[error("Unexpected end of group")]
+    UnexpectedEndOfGroup,
+
+    #[error("Unexpected start of group")]
+    UnexpectedStartOfGroup,
+}
+
+impl C8yJsonSerializer {
```
Comment (makr11st) [inline mapper/cumulocity/c8y_translator_lib/src/serializer.rs:38]: Tests are missing for `C8yJsonSerializer`.

**Review by makr11st — COMMENTED** [review]
Main thing is that we have `CumulocityJson` struct which does nothing else but holds a couple of associated methods, I think this should change.

File: mapper/cumulocity/c8y_translator_lib/src/serializer.rs (hunk 6 above)
Comment (rina23q) [reply mapper/cumulocity/c8y_translator_lib/src/serializer.rs:24]: I assume `MeasurementCollectorError` should be accurate.

File: mapper/cumulocity/c8y_translator_lib/src/serializer.rs (hunk 5 above)
Comment (rina23q) [reply mapper/cumulocity/c8y_translator_lib/src/serializer.rs:37]: I like they are separated. The highest hierarchy is `C8yJsonSerializationError`, then we have a set of collection errors under the top hierarchy.

**Review by rina23q — APPROVED** [review]
For me, it's ok.

**Review by mneumann — APPROVED** [review]
This looks okay. There are a couple of things that can be improved, but I don't want to block Lukasz on his PR.

File: mapper/cumulocity/c8y_translator_lib/src/json.rs [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+        let timestamp = local_time_now.with_timezone(local_time_now.offset());
+        let c8y_vec = Self::from_thin_edge_json_with_timestamp(input, timestamp)?;
+        Ok(c8y_vec)
+    }
+
+    fn from_thin_edge_json_with_timestamp(
+        input: &[u8],
+        timestamp: DateTime<FixedOffset>,
+    ) -> Result<Vec<u8>, serializer::C8yJsonSerializationError> {
+        let measurements = ThinEdgeJson::from_utf8(input, timestamp)?;
```
[the remark this answers is not in the document]
Comment (didier-wenzek) [reply mapper/cumulocity/c8y_translator_lib/src/json.rs:40]: I believe the argument should be renamed `default_timestamp`.

* Indeed, no timestamp is required in the input, but one is for the output.
* As done by the `from_thin_edge_json()` method, the current time is used in the output if none is provided in the input.
* A default timestamp is passed here as an argument to ease testing.

**Pull request merged**